### DISCUSSION TITLE
Kanidm's SSH CA Capabilities: Built-In Reality or a Myth?

### DISCUSSION THREAD

**Discussion by Underknowledge** [kanidm/kanidm#2831] [Q&A] [answered]
Heya, 
As I sit here daydreaming about "Warme Wurschtsemmel" and haven’t even set up Kanidm yet, but I'm already diving into its potential features. 

Has anyone here integrated SSH Certificate Authority functionalities with Kanidm? 
I remember seeing references to a built-in SSH CA feature within Kanidm, but I haven’t been able to find recent information about it. 
Could it have been deprecated, or am I possibly confusing it with another feature?

Alternatively, I am considering setting up Kanidm as an OAuth provider to facilitate authentication with Step-CA. If anyone has experience with this set-up or if there's a simpler method to achieve fancy SSH authentication, your insights would be greatly appreciated.
Cheers Friends!

**Comment by yaleman** [comment, marked as answer]
SSH Certificate Authority functionalities are still in the planning stages.

**Reply by Underknowledge** [reply]
Heya, thanks for the answer. 
I'm a bit undecided if kanidm already solves the whole authorisation part and I just go for a harder live trying to implement the CA part, The known hosts keys would be a real benefit thou.  

P.s. Thanks for your Blog, this jost pointed me towards kanidm

**Reply by Firstyear** [reply]
Known host key sharing is a really effective way to lock yourself out of a system. It was a big footgun in freeipa and we don't want to reproduce that.

Currently we went down the ssh key *distribution* path where users can upload and have ssh keys distributed to hosts. It fufils a lot of the needs, but it's not a complete replacement to ssh CA. Saying this, it is on our todo, but it's a while away. We just have other priorities right now.

**Reply by Underknowledge** [reply]
Oh no, you got this a bit wrong what I mean with the known hosts. 

On each server you integrate the CA
```
HostCertificate /etc/ssh/ssh_host_ecdsa_key-cert.pub
```  

then on connecting clients in known_hosts. 
```
@cert-authority *.example.com ecdsa-sha2-nistp256 AAAAE...=
``` 
so when you get prompted that something changed, you can say that something is not quite right. 
[source](https://smallstep.com/blog/use-ssh-certificates/) (quite the good read)
Principals and forced command "feel" like the right way. 

> We just have other priorities right now. 

yea sure! no stress, I really just seen it in a talk, but was not able to find it in the documentation. 
Its a nice to have, and as said, I'm really not sure if I really need or even want a CA. Thats something I can say after using and deploying it to multiple hosts. (Insights or opinions welcome ;) )

**Reply by Firstyear** [reply]
Ahhhh yes, that is interesting :) 

I have a longer write up on the topic over here: https://fy.blackhats.net.au/blog/2023-10-24-ssh-key-storage-comparisons/

The thing is SSH CA's solve some problems like key lifecycle, command limits etc. But you can also protect you keys in hardware with yubikeys or PIV to prevent exposure. It's not that one is better or worse, just different trades.
